File: apps/helper/src/rename.test.ts
import { describe, expect, it } from 'vitest'

import { formatEpisodeName, formatSavePath, planEpisodeRenames } from './rename'

describe('formatEpisodeName', () => {
  it('formats title - SxxExx with zero-padded season and episode', () => {
    expect(formatEpisodeName('葬送的芙莉莲', 1, 28)).toBe(
      '葬送的芙莉莲 - S01E28',
    )
    expect(formatEpisodeName('Example Show', 2, 7)).toBe(
      'Example Show - S02E07',
    )
  })
})

describe('formatSavePath', () => {
  it('joins libraryRoot / title / Season XX', () => {
    expect(formatSavePath('/library', '葬送的芙莉莲', 1)).toBe(
      '/library/葬送的芙莉莲/Season 01',
    )
    expect(formatSavePath('/library', 'Example Show', 2)).toBe(
      '/library/Example Show/Season 02',
    )
  })
})

describe('planEpisodeRenames', () => {
  it('renames the main video and matching subtitle, keeping language suffixes', () => {
    expect(
      planEpisodeRenames({
        displayName: '葬送的芙莉莲 - S01E28',
        files: [
          { name: '[ANi] 葬送的芙莉莲 - 28 [1080P].mp4', size: 700_000_000 },
          { name: '[ANi] 葬送的芙莉莲 - 28 [1080P].cht.ass', size: 40_000 },
        ],
      }),
    ).toEqual([
      {
        from: '[ANi] 葬送的芙莉莲 - 28 [1080P].mp4',
        to: '葬送的芙莉莲 - S01E28.mp4',
      },
      {
        from: '[ANi] 葬送的芙莉莲 - 28 [1080P].cht.ass',
        to: '葬送的芙莉莲 - S01E28.cht.ass',
      },
    ])
  })

  it('skips Sample, NCOP, and NCED files', () => {
    expect(
      planEpisodeRenames({
        displayName: 'Show - S01E01',
        files: [
          { name: 'Show/Show - 01.mkv', size: 1_000_000_000 },
          { name: 'Show/Sample/Show - 01 Sample.mkv', size: 20_000_000 },
          { name: 'Show/NCOP/Show NCOP.mkv', size: 50_000_000 },
          { name: 'Show/NCED/Show NCED.mkv', size: 50_000_000 },
          { name: 'Show/Show - 01.ass', size: 30_000 },
        ],
      }),
    ).toEqual([
      { from: 'Show/Show - 01.mkv', to: 'Show/Show - S01E01.mkv' },
      { from: 'Show/Show - 01.ass', to: 'Show/Show - S01E01.ass' },
    ])
  })

  it('does not plan a rename when names already match', () => {
    expect(
      planEpisodeRenames({
        displayName: 'Show - S01E01',
        files: [{ name: 'Show - S01E01.mkv', size: 1 }],
      }),
    ).toEqual([])
  })

  it('picks the largest non-extra video as the main file', () => {
    expect(
      planEpisodeRenames({
        displayName: 'Show - S01E01',
        files: [
          { name: 'Show/Show - 01.mkv', size: 2_000_000_000 },
          { name: 'Show/Show - 01 extra commentary.mp4', size: 80_000_000 },
          { name: 'Show/Show - 01.srt', size: 20_000 },
        ],
      }),
    ).toEqual([
      { from: 'Show/Show - 01.mkv', to: 'Show/Show - S01E01.mkv' },
      { from: 'Show/Show - 01.srt', to: 'Show/Show - S01E01.srt' },
    ])
  })
})

File: apps/helper-go/internal/rename/rename_test.go
package rename_test

import (
	"reflect"
	"testing"

	"github.com/Torrent-Vibe/Torrent-Vibe/apps/helper-go/internal/qb"
	"github.com/Torrent-Vibe/Torrent-Vibe/apps/helper-go/internal/rename"
)

func TestSanitizeTitleAndVideo(t *testing.T) {
	if got := rename.SanitizeTitle(`A:B`); got != "A_B" {
		t.Fatal(got)
	}
	if !rename.IsVideo("a.mkv") || !rename.IsSkippedExtra("foo.sample.mkv") {
		t.Fatal("video/extra helpers")
	}
}

func TestFormatEpisodeName(t *testing.T) {
	if got := rename.FormatEpisodeName("葬送的芙莉莲", 1, 28); got != "葬送的芙莉莲 - S01E28" {
		t.Fatal(got)
	}
	if got := rename.FormatEpisodeName("Example Show", 2, 7); got != "Example Show - S02E07" {
		t.Fatal(got)
	}
}

func TestFormatSavePath(t *testing.T) {
	if got := rename.FormatSavePath("/library", "葬送的芙莉莲", 1); got != "/library/葬送的芙莉莲/Season 01" {
		t.Fatal(got)
	}
	if got := rename.FormatSavePath("/library", "Example Show", 2); got != "/library/Example Show/Season 02" {
		t.Fatal(got)
	}
}

func TestPlanRenamesVideoAndSub(t *testing.T) {
	got := rename.PlanEpisodeRenames("葬送的芙莉莲 - S01E28", []qb.File{
		{Name: "[ANi] 葬送的芙莉莲 - 28 [1080P].mp4", Size: 700_000_000},
		{Name: "[ANi] 葬送的芙莉莲 - 28 [1080P].cht.ass", Size: 40_000},
	})
	want := []rename.Plan{
		{From: "[ANi] 葬送的芙莉莲 - 28 [1080P].mp4", To: "葬送的芙莉莲 - S01E28.mp4"},
		{From: "[ANi] 葬送的芙莉莲 - 28 [1080P].cht.ass", To: "葬送的芙莉莲 - S01E28.cht.ass"},
	}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("%+v", got)
	}
}

func TestPlanRenamesSkipsExtras(t *testing.T) {
	got := rename.PlanEpisodeRenames("Show - S01E01", []qb.File{
		{Name: "Show/Show - 01.mkv", Size: 1_000_000_000},
		{Name: "Show/Sample/Show - 01 Sample.mkv", Size: 20_000_000},
		{Name: "Show/NCOP/Show NCOP.mkv", Size: 50_000_000},
		{Name: "Show/NCED/Show NCED.mkv", Size: 50_000_000},
		{Name: "Show/Show - 01.ass", Size: 30_000},
	})
	want := []rename.Plan{
		{From: "Show/Show - 01.mkv", To: "Show/Show - S01E01.mkv"},
		{From: "Show/Show - 01.ass", To: "Show/Show - S01E01.ass"},
	}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("%+v", got)
	}
}

func TestPlanRenamesAlreadyMatch(t *testing.T) {
	got := rename.PlanEpisodeRenames("Show - S01E01", []qb.File{{Name: "Show - S01E01.mkv", Size: 1}})
	if len(got) != 0 {
		t.Fatalf("%+v", got)
	}
}

func TestPlanRenamesLargestVideo(t *testing.T) {
	got := rename.PlanEpisodeRenames("Show - S01E01", []qb.File{
		{Name: "Show/Show - 01.mkv", Size: 2_000_000_000},
		{Name: "Show/Show - 01 extra commentary.mp4", Size: 80_000_000},
		{Name: "Show/Show - 01.srt", Size: 20_000},
	})
	want := []rename.Plan{
		{From: "Show/Show - 01.mkv", To: "Show/Show - S01E01.mkv"},
		{From: "Show/Show - 01.srt", To: "Show/Show - S01E01.srt"},
	}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("%+v", got)
	}
}
